Add per-project browser config processing

BTMorton opened the pull request:
I finally got a chance to take another look at this (#71). Have hopefully addressed all of the initial feedback.

The browser helper classes are now all standalone, implementing the interface rather than extending an abstract base class.

It no longer executes the karma config file during extension configuration, but tries to figure out which browsers/custom launchers are supported by running regexp against the raw file as a string, as you've done in other places. The loader will then try to find a matching launch config from the user's final karma config when running the server and try to add the debug port flag. Unless a browser or custom launcher is set in the user's settings which will skip that part. I had to add a new env var to achieve this, though.

I've tested this with my target repo, a multi-project Angular + Electron app and it seems to be working as expected. The configuration defined in the karma.conf.js files are used by default, but can still be overriden in settings. Tests in all three projects can be run and debugged without issue.

The tests still need updating, but hopefully this is a step in the right direction to getting this functionality added. Let me know your thoughts.

BTMorton committed:
Add browser-specific configuration helpers to process custom launcher and debugger configurations
Update extension config generation to load from project karma config files

BTMorton committed:
Fix chrome's debugger flag to avoid it not starting
Add user prefs to configure firefox to allow dev tools connections

BTMorton committed:
Update browser helper classes to implement an interface, rather than extend an abstract base class
Update the config helper to avoid loading the karma config during configuration
Update the config helper to look for the browser type for custom launch configs to determine if they are supported
Update the karma config loader to use custom launcher configuration from the user's karma config if a matching config was found and it isn't overriden in settings

BTMorton committed:
Fix vscode test debugging config
Add checks to ensure the custom launcher type is supported when compiling launcher configuration
Update the electron helper to extend the chrome helper
Update extension config tests to cover new browser/custom launcher config processing

BTMorton commented:
Have updated the tests now, and expanded on the existing ones a little bit. Moving into ready for review as I think it's pretty much there, at least from my expectations.

lucono commented on src/core/config/browsers/browser-helper.ts:
```diff
@@ -0,0 +1,29 @@
+import { ConfigStore } from '../config-store.js';
+
+export interface BrowserHelper {
+  supportedBrowsers: string[];
```
Perhaps this should be typed `[string, ...string[]]` to require that implementations must support at least one browser.

lucono commented on src/core/config/browsers/electron-helper.ts:
```diff
@@ -0,0 +1,46 @@
+    config: ConfigStore<ProjectConfigSetting>
+  ): CustomLauncher {
+    const configuredLauncher: CustomLauncher = customLaucher ?? {
+      base: browserType,
```
Since we don't check here if `this.isSupportedBrowser(browserType)` (like we do in the Chrome helper), what happens if it's called with an unsupported value (such as for instance, `browserType = 'firefox'`) and no value was passed for the `customLauncher` argument?

BTMorton replied:
The theory behind it was that, if using the `BrowserFactory` to get the browser helper, you would only get the helper that supports the browser type you've specified. The exception to that rule is the chrome helper which will be used as the default, and we should fallback to the chrome browser type as the default here. So this is the only helper that is expected to have a browser type that isn't supported. I can add the checks here too, but I don't know if it's super necessary

lucono replied:
Yes, please add as it would be good to have the stand-alone helpers behave correctly even independently of the `BrowserFactory`. Everything else LGTM. Thanks.

lucono commented on src/core/config/browsers/chrome-helper.ts:
```diff
@@ -0,0 +1,113 @@
+        ? true
+        : configuredContainerMode === ContainerMode.Disabled
+        ? false
+        : isDocker();
```
Perhaps this can go in a utility function since it's used by all the helper implementations.

lucono commented on src/core/config/browsers/firefox-helper.ts:
```diff
@@ -0,0 +1,115 @@
+      } as any);
+
+    if (!this.isSupportedBrowser(configuredLauncher.base)) {
+      return configuredLauncher;
```
Not sure I'm interpreting this correctly, but say I called `getCustomLauncher` with `browserType = 'Chrome'` and no value for `customLauncher`, it seems this would return an invalid 'default' Firefox launcher configuration with launcher options that are relevant for Firefox but having a `base` value of `'Chrome'`?

BTMorton replied:
Yes, if you were to do that, that would happen. The theory, though, is that if you had `browserType = 'Chrome'` the `BrowserFactory` would return the `ChromeHelper` rather than the `FirefoxHelper`. Same as the comment you put on the `ElectronHelper`. This check is more for the case where `browserType = 'Firefox'` but the `customLauncher.base` value is somehow something entirely different.

lucono commented on src/core/config/browsers/firefox-helper.ts:
```diff
@@ -0,0 +1,115 @@
+    }
+    customLaucher.flags = customLaucher.flags?.map(flag =>
+      flag.startsWith(this.debuggingPortFlag) ? `${this.debuggingPortFlag} ${debugPort}` : flag
+    );
```
For this public method, perhaps we can return a new launcher with the new debug port value instead of mutating the supplied launcher.

BTMorton replied:
I can do that, I was just copying the original implementation here

lucono reviewed (commented):
This looks great @BTMorton. I had a couple of comments after which we can look to merge it. Thank you!

BTMorton committed:
Update to require at least one supported browser for each browser helper
Update to pass config values to the browser helper rather than the whole config store
Update to clone the custom launcher when adding debug configuration, rather than modifying in place

BTMorton committed:
Update to check incoming browser type when creating a default launcher config in all helpers
Update to check if the custom launcher base is supported as the first step in all custom launchers

lucono commented on src/frameworks/karma/config/karma-config-loader.ts:
```diff
@@ -67,10 +65,18 @@ export class KarmaConfigLoader {
+
+      const browserHelper = BrowserHelperFactory.getBrowserHelper(customLaucherObject?.base ?? '');
+
+      if (!userSpecifiedLaunchConfig) {
+        customLaucherObject = this.findMatchingCustomLauncherFromConfig(customLaucherObject, config);
+      }
```
@BTMorton Why do we try to re-match this on the Karma process side? I imagine it could discard parts of the setup performed on the extension side, such as the headless/non-headless browser setup?

BTMorton replied:
Since it no longer tries to process the karma config file on the extension side, it's much more challenging to pull out all of the custom launcher configuration. In order to make use of the user's launch config it needs to figure out if one is being used here. This puts an existing launch config slightly higher priority than the generated launch config as the user's config can be more specific. If the user specifies a browser or defines custom launch config in the extension settings, this step will be ignored. It was the best solution I could come up with at the time that achieved the goal of allowing custom launch configuration from the karma config file to be used by the extension.

lucono replied:
To use a custom launch defined in the Karma config, I imagine you only need the name of that custom launcher, which you can place in the `browsers` prop of the config on the Karma process side. It seems your implementation already made a decent effort to obtain that from the extension process side of things and I'm pretty happy with the best effort that it makes there.

BTMorton replied:
It checks if one of the custom launchers uses a supported browser and then pulls out that browser to setup debug configuration. Could make it pass through the custom launcher name as if it were the browser setting, but then that doesn't support the debug port assignments if you have multiple projects active. Would still ignore the headless/non-headless browser setup if that was configured, assuming you take the browser name first.

lucono commented on src/core/config/config-helper.ts:
```diff
@@ -48,34 +50,120 @@ export const getDefaultDebugPort = (
+ */
+export const getCustomLaunchConfiguration = (
+  config: ConfigStore<ProjectConfigSetting>,
+  projectKarmaConfigFilePath: string,
```
This should be `string | undefined` as starting with Angular 15, there isn't always a karma config file for each sub-project.

BTMorton replied:
Could this fall back to the global karma config file path somehow?

lucono replied:
I'm not familiar with a global Karma config, but I believe Angular 15+ uses an in-memory karma config, which the extension handles with [karma.conf-default.ts](https://github.com/lucono/karma-test-explorer/blob/master/src/frameworks/karma/config/karma.conf-default.ts). In the case of this update, I think it just needs to accommodate that the file can be absent and work with defaults in that scenario.

BTMorton replied:
Ah, fair enough. I was thinking there was some kind of inheritance with library projects. That makes sense, though :+1:

lucono commented on src/core/config/config-helper.ts:
```diff
@@ -48,34 +50,120 @@ export const getDefaultDebugPort = (
+  customLauncher: CustomLauncher | undefined;
+  userOverride: boolean;
+} => {
+  const rawConfig = getRawKarmaConfig(projectKarmaConfigFilePath, fileHandler);
```
In testing Angular 15+ scenarios where the Karma config file is not present on disk, this throws an exception trying to read a `null` file.

lucono commented on src/core/config/browsers/firefox-helper.ts:
```diff
@@ -0,0 +1,112 @@
+
+  public getDefaultDebuggerConfig(): DebugConfiguration {
+    return {
+      name: 'Karma Test Explorer Debugging',
+      type: this.debuggerType,
+      request: 'attach',
+      browserAttachLocation: 'workspace',
+      address: 'localhost',
+      port: FirefoxBrowserHelper.DEFAULT_DEBUGGING_PORT,
+      timeout: 60000
```
The `browserAttachLocation`, `address`, and `timeout` options don't seem to be valid for the Firefox debugger. For actual test debugging with breakpoint support, it also seems to require additional options.

BTMorton replied:
I will admit to not testing this with Firefox... That's my bad 😬

lucono commented on src/core/config/browsers/firefox-helper.ts:
```diff
@@ -0,0 +1,112 @@
+    return 'firefox';
+  }
+  public get debuggingPortFlag(): string {
+    return '-start-debugger-server';
```
Though this is right, it doesn't actually work for Firefox debugging because the `-no-remote` flag is [hard-coded](https://github.com/karma-runner/karma-firefox-launcher/blob/a55a7f43fd8ecdce02fb9a558ed769381add8083/index.js#L270) in the `karma-firefox-launcher` plugin.

lucono commented on src/core/config/extension-config.ts:
```diff
@@ -144,9 +144,28 @@ export class ExtensionConfig implements Disposable {
 
+    const { browserType, customLauncher, userOverride } = getCustomLaunchConfiguration(
+      configStore,
+      this.projectKarmaConfigFilePath,
```
This can be `null` in the cases previously mentioned.

lucono reviewed (approved):
@BTMorton this looks mostly good. I have a few comments and one question after testing, but those are minor enough that I can merge this and address them in a subsequent PR with other release updates. Thanks very much!
